### PR TITLE
DigiByte Support

### PR DESCRIPTION
**Pull request by SmartArray**
This PR will introduce support for DigiByte in renproject/multichain

### CODE REVIEW

**Comment by SmartArray** [comment]
There is still an issue with the chainCfg replacement. I need to drop the DigiByte-specific Deployments from the Deployments-Array, write some unit tests, and it should work out.
Will continue soon

**Comment by SmartArray** [comment]
##Integration finished.

DigiByte Test Coverage is 100%. (Additional AddressDecoder test for DigiByte Addresses was also added).

**Remarks**
`btcd/chaincfg` has a `Register` method that needs to be called for all bitcoin-compatible coins, if they want to support other address types. Therefore I propose to have an `Init` method for any go-package that implements a chain.
The DigiByte package has this now. An example can be seen in [digibyte.go](https://github.com/SmartArray/multichain/blob/master/chain/digibyte/digibyte.go#L9) called here [address_test.go](https://github.com/SmartArray/multichain/blob/master/chain/digibyte/address_test.go#L24)

File: docker/docker-compose.env [hunk 1]
```diff
@@ -36,4 +36,14 @@ export DOGECOIN_ADDRESS=mwjUmhAW68zCtgZpW5b1xD5g7MZew6xPV4
...
+
+#
+# DigiByte
+#
+
+# DigiByte Address that will receive mining rewards. Generally, this is set to an
+# address for which the private key is known by a test suite. This allows the test
+# suite access to plenty of testing funds.
+export DIGIBYTE_PK=edxsTg33sNnsXBPzBYB13aHRysups1SYoRfBkrYksDj9vkmNfdXo
+export DIGIBYTE_ADDRESS=dgbrt1qlpazxc2ghkaw35vzqz7ljf7ksjd5fm96wzckty
```
Comment (loongy) [inline docker/docker-compose.env:49]: Looks like these are using mainnet encodings. Addresses need to be for regnet, so I suggest changing these exports to: 
```suggestion
export DIGIBYTE_PK=eaeBaSNZmieYKXBwbwjUwzRazT45kVwi9Rubb13RJ1NgfDEUXXH8
export DIGIBYTE_ADDRESS=snhzjRDRwL5LD4ognxvjaP2St8G3MEExQp
```

File: docker/docker-compose.yaml [hunk 2]
```diff
@@ -27,9 +25,20 @@ services:
...
 
   ##
-  ## Dogecoin
+  ## DigiByte
   ##
+  digibyte:
+    build:
+      context: ./digibyte
+    ports:
+      - "0.0.0.0:18443:18443"
```
Comment (loongy) [inline docker/docker-compose.yaml:34]: These ports are already reserved by Bitcoin. Consider changing to:
```suggestion
      - "0.0.0.0:20443:18443"
```

File: multichain.go [hunk 3]
```diff
@@ -11,6 +11,7 @@ type Asset string
 const (
 	BCH  = Asset("BCH")  // Bitcoin Cash
 	BTC  = Asset("BTC")  // Bitcoin
+	DGB = Asset("DGB") // DigiByte
```
Comment (loongy) [inline multichain.go:14]: Formatting:
```suggestion
	DGB =  Asset("DGB")  // DigiByte
```

File: chain/digibyte/digibyte_test.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+				// PKH
+				pkhAddr, err := btcutil.NewAddressPubKeyHash(btcutil.Hash160(wif.PrivKey.PubKey().SerializeCompressed()), digibyte.DigiByteRegtestParams)
+				Expect(err).ToNot(HaveOccurred())
+				pkhAddrUncompressed, err := btcutil.NewAddressPubKeyHash(btcutil.Hash160(wif.PrivKey.PubKey().SerializeUncompressed()), digibyte.DigiByteRegtestParams)
+				Expect(err).ToNot(HaveOccurred())
+				log.Printf("PKH                %v", pkhAddr.EncodeAddress())
+				log.Printf("PKH (uncompressed) %v", pkhAddrUncompressed.EncodeAddress())
+
+				// Setup the client and load the unspent transaction outputs.
+				client := bitcoincompat.NewClient(bitcoincompat.DefaultClientOptions().WithHost("http://127.0.0.1:18443"))
```
Comment (loongy) [inline chain/digibyte/digibyte_test.go:43]: To match the suggestion of changing DigiByte port to `20443` (Bitcoin has already reserved `18443`):
```suggestion
				client := bitcoincompat.NewClient(bitcoincompat.DefaultClientOptions().WithHost("http://127.0.0.1:20443"))
```

**Comment by SmartArray** [comment]
Thanks for merging!

Just one thing

```
export DIGIBYTE_PK=edxsTg33sNnsXBPzBYB13aHRysups1SYoRfBkrYksDj9vkmNfdXo
export DIGIBYTE_ADDRESS=dgbrt1qlpazxc2ghkaw35vzqz7ljf7ksjd5fm96wzckty
```

This address is a regtest address!!
`dgb1` is mainnet, `dgbrt1` is regtest.

You can use the keygen.go file to generate addresses for regtest.
If you want to generate a mainnet address, just uncomment the commented line in the keygen.go file!

Thanks again 😊

**Comment by loongy** [comment]
@SmartArray that's interesting, it refused to load any UTXOs when that address when I used it.

**Comment by SmartArray** [comment]
I see. Weird that the digibyte_test suite went so well one my device? 
It definitely needs some investigation.

image

**Pull request merged**